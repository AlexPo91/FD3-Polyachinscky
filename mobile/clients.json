[ 
    {"id":101, "surname":"Иванов", "name":"Иван", "patronymic":"Иванов", "balance":200}, 
    {"id":105, "surname":"Сидоров", "name":"Сидр", "patronymic":"Сидорович", "balance":250}, 
    {"id":110, "surname":"Петров", "name":"Петр", "patronymic":"Петрович", "balance":180},
    {"id":120, "surname":"Григорьев", "name":"Григорий", "patronymic":"Григорьевич", "balance":-220}
  ]