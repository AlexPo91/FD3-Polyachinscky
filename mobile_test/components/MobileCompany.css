#root{
    display: flex;
    justify-content: center;
}
.MobileCompany {
    border: 2px solid grey;
    width: 800px;
}
.MobileCompany__item{
    padding: 20px;
    border-top: 1px solid grey;
}

.MobileCompany__item:first-child{
    border-top: none;
}
td{
    padding: 5px 10px;
}
