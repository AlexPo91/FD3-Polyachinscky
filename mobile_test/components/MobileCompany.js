import React from 'react';
import PropTypes from 'prop-types';
import MobileClient from './MobileClient';
import mobileEvents from '../events';
import FormClient from './FormClient';
import './MobileCompany.css';

class MobileCompany extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      clients: props.clients,
      workmode: 0, // 0 - кнокпа добавить, 1 - режим редактирования, 2 - режим добавления
      editableClient: null,
      filter: 'all', // all, active, blocked
    };
  }

  componentDidMount = () => {
    mobileEvents.addListener('EDeleteClient', this.deleteClient);
    mobileEvents.addListener('EEditedClient', this.editClient);
    mobileEvents.addListener('ESaveClient', this.saveClient);
    mobileEvents.addListener('ECancelEditClient', this.cancelEditClient);
  };

  componentWillUnmount = () => {
    mobileEvents.removeListener('EDeleteClient', this.deleteClient);
    mobileEvents.removeListener('EEditedClient', this.editClient);
    mobileEvents.removeListener('ESaveClient', this.saveClient);
    mobileEvents.removeListener('ECancelEditClient', this.cancelEditClient);
  };

  deleteClient = (code) => {
    const { clients } = this.state;
    // let newClients = [...clients];
    // const index = newClients.findIndex((el) => el.id === code);
    // newClients = [
    //   ...newClients.slice(0, index),
    //   ...newClients.slice(index + 1),
    // ];

    const newClients = clients.filter((el) => el.id !== code);
    this.setState({ clients: newClients });
  }

  editClient = (code) => {
    const { clients } = this.state;
    this.setState({
      workmode: 1,
      editableClient: clients.find((el) => el.id === code),
    });
  }

  addedClient =() => {
    this.setState({
      workmode: 2,
    });
  }

  cancelEditClient = () => {
    this.setState({
      workmode: 0,
    });
  }

  OnFilteredClient = (filter) => {
    this.setState({
      filter,
    });
  }

  filterClient = (clients, filter) => {
    switch (filter) {
      case 'all':
        return clients;
      case 'active':
        return clients.filter((el) => el.balance > 0);
      case 'blocked':
        return clients.filter((el) => el.balance < 0);
      default:
        return clients;
    }
  }

dataComparison = (obj1, obj2) => {
  if (obj1.surname !== obj2.surname
    || obj1.name !== obj2.name
    || obj1.patronymic !== obj2.patronymic
    || obj1.balance !== obj2.balance
  ) {
    return true;
  }
  return false;
}

  saveClient = (client) => {
    let changed = false;

    const { clients, workmode } = this.state;

    let newClients = [];
    if (workmode === 1) {
      newClients = [...clients];
      newClients.forEach((el, i) => {
        if (el.id === client.id
          && this.dataComparison(el, client)
        ) {
          changed = true;
          newClients[i] = client;
        }
      });
    } else {
      newClients = [...clients, client];
    }
    if (changed || workmode === 2) {
      this.setState({
        clients: newClients,
        workmode: 0,
      });
    }
    if (!changed) {
      this.setState({
        workmode: 0,
      });
    }
  }

  render() {
    const {
      clients, workmode, editableClient, filter,
    } = this.state;
    const visibleClients = this.filterClient(clients, filter);
    const clientsCode = visibleClients.map((el) => <MobileClient key={el.id} client={el} />);
    const uniqId = clients.length === 0 ? 10
      : clients
        .reduce((prev, current) => (prev.b > current.b ? prev : current), {}).id + 10;
    return (
      <div className="MobileCompany">
        <div className="MobileCompany__item">
          <input type="button" value="Все" onClick={() => this.OnFilteredClient('all')} />
          <input type="button" value="Активные" onClick={() => this.OnFilteredClient('active')} />
          <input type="button" value="Заблокированные" onClick={() => this.OnFilteredClient('blocked')} />
        </div>
        <div className="MobileCompany__item">
          <table>
            <thead>
              <tr>
                <th>Фамилия</th>
                <th>Имя</th>
                <th>Отчество</th>
                <th>Баланс</th>
                <th>Статус</th>
                <th>Редактировать</th>
                <th>Удалить</th>
              </tr>
            </thead>
            <tbody>
              {clientsCode}
            </tbody>
          </table>
          {workmode === 0 && <input className="addClientBtn" type="button" value="Добавить клиента" onClick={this.addedClient} />}
        </div>
        {workmode !== 0
        && (
        <FormClient
          editableClient={editableClient}
          key={workmode === 1 ? editableClient.id : null}
          workmode={workmode}
          uniqId={uniqId}
        />
        )}
      </div>
    );
  }
}

MobileCompany.propTypes = {
  clients: PropTypes.arrayOf(PropTypes.object).isRequired,
};

export default MobileCompany;
