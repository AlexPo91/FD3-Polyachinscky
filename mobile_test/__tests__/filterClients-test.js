import React from 'react';
import renderer from 'react-test-renderer';
import clients from '../clients.json';
import MobileCompany from '../components/MobileCompany';

test('работа Test Client', () => {
  const component = renderer.create(
    <MobileCompany clients={clients} />,
  );

  let componentTree = component.toJSON();
  expect(componentTree).toMatchSnapshot();

  const btnFilterActive = component.root.find((n) => n.props.value === 'Активные');
  const btnFilterBlocked = component.root.find((n) => n.props.value === 'Заблокированные');
  const btnFilterAll = component.root.find((n) => n.props.value === 'Все');

  let rowsClients = component.root.findAll((n) => n.props.className === 'trClient');

  expect(rowsClients.length).toBe(4);

  btnFilterActive.props.onClick();

  rowsClients = component.root.findAll((n) => n.props.className === 'trClient');

  expect(rowsClients.length).toBe(3);

  componentTree = component.toJSON();
  expect(componentTree).toMatchSnapshot();

  btnFilterBlocked.props.onClick();

  rowsClients = component.root.findAll((n) => n.props.className === 'trClient');

  expect(rowsClients.length).toBe(1);

  componentTree = component.toJSON();
  expect(componentTree).toMatchSnapshot();

  btnFilterAll.props.onClick();

  rowsClients = component.root.findAll((n) => n.props.className === 'trClient');

  expect(rowsClients.length).toBe(4);

  componentTree = component.toJSON();
  expect(componentTree).toMatchSnapshot();
});
