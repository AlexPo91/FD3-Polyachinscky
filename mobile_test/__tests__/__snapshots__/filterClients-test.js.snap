// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`работа Test Client 1`] = `
<div
  className="MobileCompany"
>
  <div
    className="MobileCompany__item"
  >
    <input
      onClick={[Function]}
      type="button"
      value="Все"
    />
    <input
      onClick={[Function]}
      type="button"
      value="Активные"
    />
    <input
      onClick={[Function]}
      type="button"
      value="Заблокированные"
    />
  </div>
  <div
    className="MobileCompany__item"
  >
    <table>
      <thead>
        <tr>
          <th>
            Фамилия
          </th>
          <th>
            Имя
          </th>
          <th>
            Отчество
          </th>
          <th>
            Баланс
          </th>
          <th>
            Статус
          </th>
          <th>
            Редактировать
          </th>
          <th>
            Удалить
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          className="trClient"
        >
          <td>
            Иванов
          </td>
          <td>
            Иван
          </td>
          <td>
            Иванов
          </td>
          <td
            className="tdBalance"
          >
            200
          </td>
          <td
            className="tdStatus"
            style={
              Object {
                "backgroundColor": "green",
              }
            }
          >
            active
          </td>
          <td
            className="tdEditBtn"
          >
            <input
              className="btnEdit"
              onClick={[Function]}
              type="button"
              value="Редактировать"
            />
          </td>
          <td
            className="tdDeleteBtn"
          >
            <input
              className="btnDelete"
              onClick={[Function]}
              type="button"
              value="Удалить"
            />
          </td>
        </tr>
        <tr
          className="trClient"
        >
          <td>
            Сидоров
          </td>
          <td>
            Сидр
          </td>
          <td>
            Сидорович
          </td>
          <td
            className="tdBalance"
          >
            250
          </td>
          <td
            className="tdStatus"
            style={
              Object {
                "backgroundColor": "green",
              }
            }
          >
            active
          </td>
          <td
            className="tdEditBtn"
          >
            <input
              className="btnEdit"
              onClick={[Function]}
              type="button"
              value="Редактировать"
            />
          </td>
          <td
            className="tdDeleteBtn"
          >
            <input
              className="btnDelete"
              onClick={[Function]}
              type="button"
              value="Удалить"
            />
          </td>
        </tr>
        <tr
          className="trClient"
        >
          <td>
            Петров
          </td>
          <td>
            Петр
          </td>
          <td>
            Петрович
          </td>
          <td
            className="tdBalance"
          >
            1
          </td>
          <td
            className="tdStatus"
            style={
              Object {
                "backgroundColor": "green",
              }
            }
          >
            active
          </td>
          <td
            className="tdEditBtn"
          >
            <input
              className="btnEdit"
              onClick={[Function]}
              type="button"
              value="Редактировать"
            />
          </td>
          <td
            className="tdDeleteBtn"
          >
            <input
              className="btnDelete"
              onClick={[Function]}
              type="button"
              value="Удалить"
            />
          </td>
        </tr>
        <tr
          className="trClient"
        >
          <td>
            Григорьев
          </td>
          <td>
            Григорий
          </td>
          <td>
            Григорьевич
          </td>
          <td
            className="tdBalance"
          >
            -220
          </td>
          <td
            className="tdStatus"
            style={
              Object {
                "backgroundColor": "red",
              }
            }
          >
            blocked
          </td>
          <td
            className="tdEditBtn"
          >
            <input
              className="btnEdit"
              onClick={[Function]}
              type="button"
              value="Редактировать"
            />
          </td>
          <td
            className="tdDeleteBtn"
          >
            <input
              className="btnDelete"
              onClick={[Function]}
              type="button"
              value="Удалить"
            />
          </td>
        </tr>
      </tbody>
    </table>
    <input
      className="addClientBtn"
      onClick={[Function]}
      type="button"
      value="Добавить клиента"
    />
  </div>
</div>
`;

exports[`работа Test Client 2`] = `
<div
  className="MobileCompany"
>
  <div
    className="MobileCompany__item"
  >
    <input
      onClick={[Function]}
      type="button"
      value="Все"
    />
    <input
      onClick={[Function]}
      type="button"
      value="Активные"
    />
    <input
      onClick={[Function]}
      type="button"
      value="Заблокированные"
    />
  </div>
  <div
    className="MobileCompany__item"
  >
    <table>
      <thead>
        <tr>
          <th>
            Фамилия
          </th>
          <th>
            Имя
          </th>
          <th>
            Отчество
          </th>
          <th>
            Баланс
          </th>
          <th>
            Статус
          </th>
          <th>
            Редактировать
          </th>
          <th>
            Удалить
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          className="trClient"
        >
          <td>
            Иванов
          </td>
          <td>
            Иван
          </td>
          <td>
            Иванов
          </td>
          <td
            className="tdBalance"
          >
            200
          </td>
          <td
            className="tdStatus"
            style={
              Object {
                "backgroundColor": "green",
              }
            }
          >
            active
          </td>
          <td
            className="tdEditBtn"
          >
            <input
              className="btnEdit"
              onClick={[Function]}
              type="button"
              value="Редактировать"
            />
          </td>
          <td
            className="tdDeleteBtn"
          >
            <input
              className="btnDelete"
              onClick={[Function]}
              type="button"
              value="Удалить"
            />
          </td>
        </tr>
        <tr
          className="trClient"
        >
          <td>
            Сидоров
          </td>
          <td>
            Сидр
          </td>
          <td>
            Сидорович
          </td>
          <td
            className="tdBalance"
          >
            250
          </td>
          <td
            className="tdStatus"
            style={
              Object {
                "backgroundColor": "green",
              }
            }
          >
            active
          </td>
          <td
            className="tdEditBtn"
          >
            <input
              className="btnEdit"
              onClick={[Function]}
              type="button"
              value="Редактировать"
            />
          </td>
          <td
            className="tdDeleteBtn"
          >
            <input
              className="btnDelete"
              onClick={[Function]}
              type="button"
              value="Удалить"
            />
          </td>
        </tr>
        <tr
          className="trClient"
        >
          <td>
            Петров
          </td>
          <td>
            Петр
          </td>
          <td>
            Петрович
          </td>
          <td
            className="tdBalance"
          >
            1
          </td>
          <td
            className="tdStatus"
            style={
              Object {
                "backgroundColor": "green",
              }
            }
          >
            active
          </td>
          <td
            className="tdEditBtn"
          >
            <input
              className="btnEdit"
              onClick={[Function]}
              type="button"
              value="Редактировать"
            />
          </td>
          <td
            className="tdDeleteBtn"
          >
            <input
              className="btnDelete"
              onClick={[Function]}
              type="button"
              value="Удалить"
            />
          </td>
        </tr>
      </tbody>
    </table>
    <input
      className="addClientBtn"
      onClick={[Function]}
      type="button"
      value="Добавить клиента"
    />
  </div>
</div>
`;

exports[`работа Test Client 3`] = `
<div
  className="MobileCompany"
>
  <div
    className="MobileCompany__item"
  >
    <input
      onClick={[Function]}
      type="button"
      value="Все"
    />
    <input
      onClick={[Function]}
      type="button"
      value="Активные"
    />
    <input
      onClick={[Function]}
      type="button"
      value="Заблокированные"
    />
  </div>
  <div
    className="MobileCompany__item"
  >
    <table>
      <thead>
        <tr>
          <th>
            Фамилия
          </th>
          <th>
            Имя
          </th>
          <th>
            Отчество
          </th>
          <th>
            Баланс
          </th>
          <th>
            Статус
          </th>
          <th>
            Редактировать
          </th>
          <th>
            Удалить
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          className="trClient"
        >
          <td>
            Григорьев
          </td>
          <td>
            Григорий
          </td>
          <td>
            Григорьевич
          </td>
          <td
            className="tdBalance"
          >
            -220
          </td>
          <td
            className="tdStatus"
            style={
              Object {
                "backgroundColor": "red",
              }
            }
          >
            blocked
          </td>
          <td
            className="tdEditBtn"
          >
            <input
              className="btnEdit"
              onClick={[Function]}
              type="button"
              value="Редактировать"
            />
          </td>
          <td
            className="tdDeleteBtn"
          >
            <input
              className="btnDelete"
              onClick={[Function]}
              type="button"
              value="Удалить"
            />
          </td>
        </tr>
      </tbody>
    </table>
    <input
      className="addClientBtn"
      onClick={[Function]}
      type="button"
      value="Добавить клиента"
    />
  </div>
</div>
`;

exports[`работа Test Client 4`] = `
<div
  className="MobileCompany"
>
  <div
    className="MobileCompany__item"
  >
    <input
      onClick={[Function]}
      type="button"
      value="Все"
    />
    <input
      onClick={[Function]}
      type="button"
      value="Активные"
    />
    <input
      onClick={[Function]}
      type="button"
      value="Заблокированные"
    />
  </div>
  <div
    className="MobileCompany__item"
  >
    <table>
      <thead>
        <tr>
          <th>
            Фамилия
          </th>
          <th>
            Имя
          </th>
          <th>
            Отчество
          </th>
          <th>
            Баланс
          </th>
          <th>
            Статус
          </th>
          <th>
            Редактировать
          </th>
          <th>
            Удалить
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          className="trClient"
        >
          <td>
            Иванов
          </td>
          <td>
            Иван
          </td>
          <td>
            Иванов
          </td>
          <td
            className="tdBalance"
          >
            200
          </td>
          <td
            className="tdStatus"
            style={
              Object {
                "backgroundColor": "green",
              }
            }
          >
            active
          </td>
          <td
            className="tdEditBtn"
          >
            <input
              className="btnEdit"
              onClick={[Function]}
              type="button"
              value="Редактировать"
            />
          </td>
          <td
            className="tdDeleteBtn"
          >
            <input
              className="btnDelete"
              onClick={[Function]}
              type="button"
              value="Удалить"
            />
          </td>
        </tr>
        <tr
          className="trClient"
        >
          <td>
            Сидоров
          </td>
          <td>
            Сидр
          </td>
          <td>
            Сидорович
          </td>
          <td
            className="tdBalance"
          >
            250
          </td>
          <td
            className="tdStatus"
            style={
              Object {
                "backgroundColor": "green",
              }
            }
          >
            active
          </td>
          <td
            className="tdEditBtn"
          >
            <input
              className="btnEdit"
              onClick={[Function]}
              type="button"
              value="Редактировать"
            />
          </td>
          <td
            className="tdDeleteBtn"
          >
            <input
              className="btnDelete"
              onClick={[Function]}
              type="button"
              value="Удалить"
            />
          </td>
        </tr>
        <tr
          className="trClient"
        >
          <td>
            Петров
          </td>
          <td>
            Петр
          </td>
          <td>
            Петрович
          </td>
          <td
            className="tdBalance"
          >
            1
          </td>
          <td
            className="tdStatus"
            style={
              Object {
                "backgroundColor": "green",
              }
            }
          >
            active
          </td>
          <td
            className="tdEditBtn"
          >
            <input
              className="btnEdit"
              onClick={[Function]}
              type="button"
              value="Редактировать"
            />
          </td>
          <td
            className="tdDeleteBtn"
          >
            <input
              className="btnDelete"
              onClick={[Function]}
              type="button"
              value="Удалить"
            />
          </td>
        </tr>
        <tr
          className="trClient"
        >
          <td>
            Григорьев
          </td>
          <td>
            Григорий
          </td>
          <td>
            Григорьевич
          </td>
          <td
            className="tdBalance"
          >
            -220
          </td>
          <td
            className="tdStatus"
            style={
              Object {
                "backgroundColor": "red",
              }
            }
          >
            blocked
          </td>
          <td
            className="tdEditBtn"
          >
            <input
              className="btnEdit"
              onClick={[Function]}
              type="button"
              value="Редактировать"
            />
          </td>
          <td
            className="tdDeleteBtn"
          >
            <input
              className="btnDelete"
              onClick={[Function]}
              type="button"
              value="Удалить"
            />
          </td>
        </tr>
      </tbody>
    </table>
    <input
      className="addClientBtn"
      onClick={[Function]}
      type="button"
      value="Добавить клиента"
    />
  </div>
</div>
`;
